internal/provider/resource_workspace_run.go

```diff
@@ -0,0 +1,359 @@
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	tharsis "gitlab.com/infor-cloud/martian-cloud/tharsis/tharsis-sdk-go/pkg"
+	sdktypes "gitlab.com/infor-cloud/martian-cloud/tharsis/tharsis-sdk-go/pkg/types"
+)
+
+const (
+	jobCompletionPollInterval = 5 * time.Second
+)
+
+var (
+	applyRunComment = "terraform-provider-tharsis" // must be var, not const, to take address
+)
+
+// WorkspaceRunModel is the model for a workspace_run.
+// Please note: Unlike many/most other resources, this model does not exist in the Tharsis API.
+// The workspace path, module source, and module version uniquely identify this workspace_run.
+type WorkspaceRunModel struct {
+	WorkspacePath types.String `tfsdk:"workspace_path"`
+	ModuleSource  types.String `tfsdk:"module_source"`
+	ModuleVersion types.String `tfsdk:"module_version"`
+}
+
+// Ensure provider defined types fully satisfy framework interfaces
+var (
+	_ resource.Resource                = (*workspaceRunResource)(nil)
+	_ resource.ResourceWithConfigure   = (*workspaceRunResource)(nil)
+	_ resource.ResourceWithImportState = (*workspaceRunResource)(nil)
+)
+
+// NewWorkspaceRunResource is a helper function to simplify the provider implementation.
+func NewWorkspaceRunResource() resource.Resource {
+	return &workspaceRunResource{}
+}
+
+type workspaceRunResource struct {
+	client *tharsis.Client
+}
+
+// Metadata returns the full name of the resource, including prefix, underscore, instance name.
+func (t *workspaceRunResource) Metadata(ctx context.Context,
+	req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = "tharsis_workspace_run"
+}
+
+func (t *workspaceRunResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	description := "Defines and manages a workspace run."
+
+	resp.Schema = schema.Schema{
+		Version:             1,
+		MarkdownDescription: description,
+		Description:         description,
+		Attributes: map[string]schema.Attribute{
+			"workspace_path": schema.StringAttribute{
+				MarkdownDescription: "The full path of the workspace.",
+				Description:         "The full path of the workspace.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"module_source": schema.StringAttribute{
+				MarkdownDescription: "The source of the module, including the API hostname.",
+				Description:         "The source of the module, including the API hostname.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
```

Comment:
We shouldn't need to destroy re-apply of the module changes. The module could have been renamed and it shouldn't matter. Even if it is completely different, terraform will destroy all resources and create the new ones.